File: node_version/github.ts
import axios, { AxiosInstance } from "axios";
import { loadConfig } from "./config.js";

const GITHUB_API_BASE = "https://api.github.com";

export type RepoLanguageMap = Record<string, number>;

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function asNonEmptyString(value: unknown): string | null {
  if (typeof value !== "string") {
    return null;
  }

  const trimmed = value.trim();
  return trimmed ? trimmed : null;
}

function getGithubToken(overrideToken?: string): string | null {
  const direct = asNonEmptyString(overrideToken);
  if (direct) {
    return direct;
  }

  try {
    const cfg = (loadConfig() ?? {}) as Record<string, any>;
    const githubCfg = cfg.github as Record<string, any> | undefined;
    const configToken = asNonEmptyString(githubCfg?.token);
    if (configToken) {
      return configToken;
    }
  } catch {
    // Ignore config read errors and fall back to env
  }

  const envToken = asNonEmptyString(
    process.env.GITHUB_TOKEN || process.env.GH_TOKEN,
  );
  if (envToken) {
    return envToken;
  }

  return null;
}

function createGithubClient(token?: string): AxiosInstance {
  const headers: Record<string, string> = {
    Accept: "application/vnd.github+json",
    "User-Agent": "explainthisrepo/1.0",
  };

  const resolvedToken = getGithubToken(token);
  if (resolvedToken) {
    headers.Authorization = `Bearer ${resolvedToken}`;
  }

  return axios.create({
    baseURL: GITHUB_API_BASE,
    headers,
    timeout: 10_000,
  });
}

function isRateLimitText(text: unknown): boolean {
  const lower = typeof text === "string" ? text.toLowerCase() : "";
  return (
    lower.includes("secondary rate limit") ||
    lower.includes("rate limit")
  );
}

function rateLimitMessage(response?: { headers?: Record<string, unknown> }): string {
  const reset = response?.headers?.["x-ratelimit-reset"];

  if (typeof reset === "string" || typeof reset === "number") {
    const resetTs = Number(reset);
    if (!Number.isNaN(resetTs)) {
      const waitSeconds = Math.max(0, resetTs * 1000 - Date.now()) / 1000;
      const mins = Math.max(1, Math.ceil(waitSeconds / 60));

      return (
        "GitHub API rate limit exceeded.\n" +
        `Try again in ~${mins} minute(s).\n` +
        "Fix:\n" +
        "- Set GITHUB_TOKEN in config or environment\n" +
        "- Or run `explainthisrepo init`\n"
      );
    }
  }

  return (
    "GitHub API rate limit exceeded.\n" +
    "Fix:\n" +
    "- Set GITHUB_TOKEN in config or environment\n" +
    "- Or run `explainthisrepo init`\n"
  );
}

function privateRepoMessage(): string {
  return (
    "Repository not found.\n" +
    "If this is a private repository, configure GitHub access:\n" +
    "- Run `explainthisrepo init`\n" +
    "- Or set GITHUB_TOKEN (see docs/GITHUB_TOKEN.md)\n"
  );
}

function formatJsonError(err: unknown): string {
  if (!axios.isAxiosError(err)) {
    return `GitHub request failed: ${String(err)}`;
  }

  const status = err.response?.status;
  const data = err.response?.data as any;
  const message = typeof data?.message === "string" ? data.message : err.message;

  if (status === 404) {
    return privateRepoMessage();
  }

  if (status === 401) {
    return "GitHub auth failed (401). Invalid GitHub token.";
  }

  if (status === 403) {
    if (
      err.response?.headers?.["x-ratelimit-remaining"] === "0" ||
      isRateLimitText(message)
    ) {
      return rateLimitMessage(err.response);
    }
    return "GitHub API access forbidden (403).";
  }

  if (status === 429) {
    return rateLimitMessage(err.response);
  }

  if (status && status >= 500) {
    return `GitHub API server error (${status}). Try again later.`;
  }

  if (status !== undefined) {
    return `GitHub API request failed (${status}): ${message}`;
  }

  return `Network error while calling GitHub: ${message}`;
}

async function requestJson<T>(
  client: AxiosInstance,
  url: string,
  opts?: { params?: Record<string, unknown>; maxRetries?: number; baseDelayMs?: number },
): Promise<T> {
  const maxRetries = opts?.maxRetries ?? 4;
  const baseDelayMs = opts?.baseDelayMs ?? 700;

  let attempt = 0;
  let backoff = baseDelayMs;

  while (attempt <= maxRetries) {
    try {
      const response = await client.get<T>(url, {
        params: opts?.params,
      });
      return response.data;
    } catch (err: unknown) {
      if (!axios.isAxiosError(err)) {
        if (attempt >= maxRetries) {
          throw new Error(`Network error while calling GitHub: ${String(err)}`);
        }
        await sleep(backoff);
        backoff *= 2;
        attempt += 1;
        continue;
      }

      const status = err.response?.status;
      const data = err.response?.data as any;
      const message = typeof data?.message === "string" ? data.message : err.message;
      const remaining = err.response?.headers?.["x-ratelimit-remaining"];

      if (status === 404) {
        throw new Error(privateRepoMessage());
      }

      if (status === 403 || status === 429) {
        const rateLimited = remaining === "0" || isRateLimitText(message);

        if (rateLimited) {
          if (attempt >= maxRetries) {
            throw new Error(rateLimitMessage(err.response));
          }
          await sleep(backoff);
          backoff *= 2;
          attempt += 1;
          continue;
        }

        if (status === 429) {
          if (attempt >= maxRetries) {
            throw new Error(rateLimitMessage(err.response));
          }
          await sleep(backoff);
          backoff *= 2;
          attempt += 1;
          continue;
        }

        throw new Error("GitHub API access forbidden (403).");
      }

      if (status && status >= 500) {
        if (attempt >= maxRetries) {
          throw new Error(
            `GitHub API server error (${status}). Try again later.`,
          );
        }
        await sleep(backoff);
        backoff *= 2;
        attempt += 1;
        continue;
      }

      if (status === undefined) {
        if (attempt >= maxRetries) {
          throw new Error(`Network error while calling GitHub: ${message}`);
        }
        await sleep(backoff);
        backoff *= 2;
        attempt += 1;
        continue;
      }

      throw new Error(formatJsonError(err));
    }
  }

  throw new Error("GitHub request failed unexpectedly.");
}

async function requestText(
  client: AxiosInstance,
  url: string,
  opts?: { accept?: string; maxRetries?: number; baseDelayMs?: number; params?: Record<string, unknown> },
): Promise<string | null> {
  const accept = opts?.accept ?? "application/vnd.github.v3.raw";
  const maxRetries = opts?.maxRetries ?? 4;
  const baseDelayMs = opts?.baseDelayMs ?? 700;

  let attempt = 0;
  let backoff = baseDelayMs;

  while (attempt <= maxRetries) {
    try {
      const response = await client.get<string>(url, {
        params: opts?.params,
        headers: { Accept: accept },
        responseType: "text",
        transformResponse: [(data) => data],
      });

      return typeof response.data === "string" ? response.data : String(response.data ?? "");
    } catch (err: unknown) {
      if (!axios.isAxiosError(err)) {
        if (attempt >= maxRetries) {
          return null;
        }
        await sleep(backoff);
        backoff *= 2;
        attempt += 1;
        continue;
      }

      const status = err.response?.status;
      const data = err.response?.data as any;
      const message = typeof data?.message === "string" ? data.message : err.message;
      const remaining = err.response?.headers?.["x-ratelimit-remaining"];

      if (status === 404) {
        return null;
      }

      if (status === 403 || status === 429) {
        const rateLimited = remaining === "0" || isRateLimitText(message);

        if (rateLimited) {
          if (attempt >= maxRetries) {
            return null;
          }
          await sleep(backoff);
          backoff *= 2;
          attempt += 1;
          continue;
        }

        return null;
      }

      if (status && status >= 500) {
        if (attempt >= maxRetries) {
          return null;
        }
        await sleep(backoff);
        backoff *= 2;
        attempt += 1;
        continue;
      }

      return null;
    }
  }

  return null;
}

export async function fetchRepo(owner: string, repo: string, token?: string): Promise<any> {
  const client = createGithubClient(token);
  return requestJson<any>(client, `/repos/${owner}/${repo}`);
}

export async function fetchReadme(
  owner: string,
  repo: string,
  token?: string,
): Promise<string | null> {
  const client = createGithubClient(token);

  const apiUrl = `/repos/${owner}/${repo}/readme`;
  const text = await requestText(client, apiUrl, {
    accept: "application/vnd.github.v3.raw",
    maxRetries: 4,
  });

  if (text) {
    return text;
  }

  const branches = ["main", "master"];
  const filenames = ["README.md", "readme.md", "README.MD"];

  for (const branch of branches) {
    for (const name of filenames) {
      const rawUrl = `https://raw.githubusercontent.com/${owner}/${repo}/${branch}/${name}`;
      const raw = await requestText(client, rawUrl, {
        accept: "text/plain",
        maxRetries: 2,
      });

      if (raw) {
        return raw;
      }
    }
  }

  return null;
}

export async function fetchTree(
  owner: string,
  repo: string,
  token?: string,
): Promise<any[]> {
  const client = createGithubClient(token);

  const repoMeta = await fetchRepo(owner, repo, token);
  const defaultBranch = repoMeta?.default_branch || "main";

  const data = await requestJson<any>(
    client,
    `/repos/${owner}/${repo}/git/trees/${defaultBranch}`,
    { params: { recursive: 1 } },
  );

  const tree = data?.tree || [];
  if (!Array.isArray(tree)) {
    return [];
  }

  // Kept in the current shape to avoid breaking downstream callers.
  return tree.map((item: any) => ({
    path: item.path,
    type: item.type,
    size: item.size,
  }));
}

export async function fetchFile(
  owner: string,
  repo: string,
  filePath: string,
  token?: string,
): Promise<string | null> {
  const client = createGithubClient(token);

  return requestText(client, `/repos/${owner}/${repo}/contents/${filePath}`, {
    accept: "application/vnd.github.v3.raw",
    maxRetries: 2,
  });
}

export async function fetchLanguages(
  owner: string,
  repo: string,
  token?: string,
): Promise<RepoLanguageMap> {
  const client = createGithubClient(token);
  return requestJson<RepoLanguageMap>(client, `/repos/${owner}/${repo}/languages`);
}

File: explain_this_repo/github.py
from __future__ import annotations

import base64
import binascii
import os
import time
from typing import Optional
from urllib.parse import quote

import requests

from explain_this_repo.config import load_config
from explain_this_repo.file_reader import FileReadResult, build_file_read_result

GITHUB_API_BASE = "https://api.github.com"
_MAX_FILE_BYTES = 32_000


def _get_token(token: Optional[str] = None) -> Optional[str]:
    if token and token.strip():
        return token.strip()

    try:
        cfg = load_config() or {}
        if isinstance(cfg, dict):
            github_cfg = cfg.get("github", {})
            if isinstance(github_cfg, dict):
                cfg_token = github_cfg.get("token")
                if cfg_token and str(cfg_token).strip():
                    return str(cfg_token).strip()
    except Exception:
        pass

    env_token = os.getenv("GITHUB_TOKEN") or os.getenv("GH_TOKEN")
    if env_token and env_token.strip():
        return env_token.strip()

    return None


def _make_session(token: Optional[str] = None) -> requests.Session:
    session = requests.Session()

    headers = {
        "User-Agent": "explainthisrepo/1.0",
        "Accept": "application/vnd.github+json",
    }

    resolved_token = _get_token(token)
    if resolved_token:
        headers["Authorization"] = f"Bearer {resolved_token}"

    session.headers.update(headers)
    return session


def _rate_limit_message(response: requests.Response) -> str:
    remaining = response.headers.get("X-RateLimit-Remaining")
    reset = response.headers.get("X-RateLimit-Reset")

    if remaining == "0" and reset:
        try:
            reset_ts = int(reset)
            wait_s = max(0, reset_ts - int(time.time()))
            mins = (wait_s + 59) // 60
            return (
                "GitHub API rate limit exceeded.\n"
                f"Try again in ~{mins} minute(s).\n"
                "Fix:\n"
                "- Set GITHUB_TOKEN in config or environment\n"
                "- Or run `explainthisrepo init`\n"
            )
        except Exception:
            pass

    return (
        "GitHub API rate limit exceeded.\n"
        "Fix:\n"
        "- Set GITHUB_TOKEN in config or environment\n"
        "- Or run `explainthisrepo init`\n"
    )


def _request_json(
    session: requests.Session,
    url: str,
    *,
    timeout: int = 10,
    retries: int = 4,
) -> dict:
    backoff = 1.5

    for attempt in range(retries + 1):
        try:
            response = session.get(url, timeout=timeout)
        except requests.RequestException as e:
            if attempt == retries:
                raise RuntimeError(f"Network error while calling GitHub: {e}") from e
            time.sleep(backoff)
            backoff *= 2
            continue

        if response.status_code == 200:
            return response.json()

        if response.status_code == 404:
            raise RuntimeError(
                "Repository not found.\n"
                "If this is a private repository, configure GitHub access:\n"
                "- Run `explainthisrepo init`\n"
                "- Or set GITHUB_TOKEN (see https://github.com/calchiwo/ExplainThisRepo/blob/main/docs/GITHUB_TOKEN.md)"
            )

        if response.status_code in (403, 429):
            text_lower = (response.text or "").lower()

            if response.headers.get("X-RateLimit-Remaining") == "0":
                raise RuntimeError(_rate_limit_message(response))

            if "secondary rate limit" in text_lower or "rate limit" in text_lower:
                if attempt == retries:
                    raise RuntimeError(_rate_limit_message(response))
                time.sleep(backoff)
                backoff *= 2
                continue

            if "resource not accessible by integration" in text_lower or "private" in text_lower:
                raise RuntimeError("GitHub API access forbidden (403).")

            if attempt == retries:
                raise RuntimeError("GitHub API access forbidden (403).")

            time.sleep(backoff)
            backoff *= 2
            continue

        if 500 <= response.status_code <= 599:
            if attempt == retries:
                raise RuntimeError(
                    f"GitHub API server error ({response.status_code})."
                )
            time.sleep(backoff)
            backoff *= 2
            continue

        raise RuntimeError(f"GitHub API request failed ({response.status_code}).")

    raise RuntimeError("GitHub request failed unexpectedly.")


def _request_text(
    session: requests.Session,
    url: str,
    *,
    accept: str,
    timeout: int = 10,
    retries: int = 4,
) -> Optional[str]:
    backoff = 1.5

    for attempt in range(retries + 1):
        try:
            response = session.get(url, headers={"Accept": accept}, timeout=timeout)
        except requests.RequestException:
            if attempt == retries:
                return None
            time.sleep(backoff)
            backoff *= 2
            continue

        if response.status_code == 200:
            return response.text

        if response.status_code == 404:
            return None

        if response.status_code in (403, 429):
            text_lower = (response.text or "").lower()

            if response.headers.get("X-RateLimit-Remaining") == "0":
                return None

            if "secondary rate limit" in text_lower or "rate limit" in text_lower:
                if attempt == retries:
                    return None
                time.sleep(backoff)
                backoff *= 2
                continue

            if "resource not accessible by integration" in text_lower or "private" in text_lower:
                return None

            if attempt == retries:
                return None

            time.sleep(backoff)
            backoff *= 2
            continue

        if 500 <= response.status_code <= 599:
            if attempt == retries:
                return None
            time.sleep(backoff)
            backoff *= 2
            continue

        return None

    return None


def _normalize_github_path(file_path: str) -> str:
    cleaned = file_path.strip()
    while cleaned.startswith("./"):
        cleaned = cleaned[2:]
    cleaned = cleaned.lstrip("/")
    if not cleaned:
        raise RuntimeError("GitHub file path is empty.")
    return cleaned


def _quote_github_path(file_path: str) -> str:
    return quote(file_path, safe="/")


def _decode_base64_content(content: str) -> bytes:
    cleaned = "".join(content.split())
    if not cleaned:
        return b""

    pad = (-len(cleaned)) % 4
    if pad:
        cleaned += "=" * pad

    try:
        return base64.b64decode(cleaned, validate=True)
    except (ValueError, binascii.Error) as e:
        raise RuntimeError("Failed to decode GitHub base64 file content.") from e


def _request_contents_json(
    session: requests.Session,
    url: str,
    *,
    owner: str,
    repo: str,
    path: str,
    timeout: int = 10,
    retries: int = 4,
) -> object:
    backoff = 1.5

    for attempt in range(retries + 1):
        try:
            response = session.get(url, timeout=timeout)
        except requests.RequestException as e:
            if attempt == retries:
                raise RuntimeError(f"Network error while calling GitHub: {e}") from e
            time.sleep(backoff)
            backoff *= 2
            continue

        if response.status_code == 200:
            return response.json()

        if response.status_code == 404:
            raise RuntimeError(f"GitHub 404: {owner}/{repo}/{path} not found.")

        if response.status_code in (403, 429):
            text_lower = (response.text or "").lower()

            if response.headers.get("X-RateLimit-Remaining") == "0":
                raise RuntimeError(_rate_limit_message(response))

            if "secondary rate limit" in text_lower or "rate limit" in text_lower:
                if attempt == retries:
                    raise RuntimeError(_rate_limit_message(response))
                time.sleep(backoff)
                backoff *= 2
                continue

            if "resource not accessible by integration" in text_lower or "private" in text_lower:
                raise RuntimeError(
                    f"GitHub API access forbidden (403) for {owner}/{repo}/{path}.\n"
                    "If this is a private repository, check your GitHub token permissions."
                )

            if attempt == retries:
                raise RuntimeError(
                    f"GitHub API access forbidden (403) for {owner}/{repo}/{path}."
                )

            time.sleep(backoff)
            backoff *= 2
            continue

        if 500 <= response.status_code <= 599:
            if attempt == retries:
                raise RuntimeError(
                    f"GitHub API server error ({response.status_code}) while fetching {owner}/{repo}/{path}."
                )
            time.sleep(backoff)
            backoff *= 2
            continue

        raise RuntimeError(
            f"GitHub API request failed ({response.status_code}) while fetching {owner}/{repo}/{path}."
        )

    raise RuntimeError("GitHub request failed unexpectedly.")


def fetch_repo(owner: str, repo: str, token: Optional[str] = None) -> dict:
    session = _make_session(token)
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}"
    return _request_json(session, url)


def fetch_readme(owner: str, repo: str, token: Optional[str] = None) -> str | None:
    session = _make_session(token)

    api_url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/readme"
    text = _request_text(
        session,
        api_url,
        accept="application/vnd.github.v3.raw",
    )
    if text:
        return text

    branches = ["main", "master"]
    filenames = ["README.md", "readme.md", "README.MD"]

    for branch in branches:
        for name in filenames:
            raw_url = f"https://raw.githubusercontent.com/{owner}/{repo}/{branch}/{name}"
            raw = _request_text(
                session,
                raw_url,
                accept="text/plain",
                timeout=10,
                retries=2,
            )
            if raw:
                return raw

    return None


def fetch_tree(owner: str, repo: str, token: Optional[str] = None) -> list[dict]:
    session = _make_session(token)
    repo_meta = fetch_repo(owner, repo, token=token)
    branch = repo_meta.get("default_branch") or "main"

    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/git/trees/{branch}?recursive=1"
    data = _request_json(session, url)

    tree = data.get("tree", [])
    if not isinstance(tree, list):
        return []
    return tree


def fetch_file(
    owner: str,
    repo: str,
    file_path: str,
    token: Optional[str] = None,
) -> str | None:
    session = _make_session(token)
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/contents/{file_path}"
    return _request_text(
        session,
        url,
        accept="application/vnd.github.v3.raw",
        timeout=10,
        retries=2,
    )


def fetch_file_result(
    owner: str,
    repo: str,
    file_path: str,
    token: Optional[str] = None,
    max_bytes: int = _MAX_FILE_BYTES,
) -> FileReadResult:
    normalized_path = _normalize_github_path(file_path)
    session = _make_session(token)
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/contents/{_quote_github_path(normalized_path)}"

    payload = _request_contents_json(
        session,
        url,
        owner=owner,
        repo=repo,
        path=normalized_path,
    )

    if isinstance(payload, list):
        raise RuntimeError(
            f"GitHub path resolves to a directory, not a file: {owner}/{repo}/{normalized_path}"
        )

    if not isinstance(payload, dict):
        raise RuntimeError(
            f"GitHub API returned unexpected data for {owner}/{repo}/{normalized_path}."
        )

    entry_type = str(payload.get("type") or "").lower()
    if entry_type != "file":
        raise RuntimeError(
            f"GitHub path does not resolve to a regular file: {owner}/{repo}/{normalized_path}"
        )

    encoding = str(payload.get("encoding") or "").lower()
    if encoding != "base64":
        raise RuntimeError(
            f"GitHub file encoding is unsupported for {owner}/{repo}/{normalized_path}."
        )

    content_field = payload.get("content")
    if not isinstance(content_field, str) or not content_field.strip():
        raise RuntimeError(
            f"GitHub file content is unavailable for {owner}/{repo}/{normalized_path}."
        )

    raw = _decode_base64_content(content_field)

    size_value = payload.get("size")
    try:
        size_bytes = int(size_value) if size_value is not None else len(raw)
    except Exception:
        size_bytes = len(raw)

    return build_file_read_result(
        path=normalized_path,
        raw=raw,
        size_bytes=size_bytes,
        max_bytes=max_bytes,
    )


def fetch_directory_contents(
    owner: str,
    repo: str,
    directory_path: str,
    token: Optional[str] = None,
) -> list[dict]:
    normalized_path = _normalize_github_path(directory_path)
    session = _make_session(token)
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/contents/{_quote_github_path(normalized_path)}"

    payload = _request_contents_json(
        session,
        url,
        owner=owner,
        repo=repo,
        path=normalized_path,
    )

    if isinstance(payload, dict):
        entry_type = str(payload.get("type") or "").lower()
        if entry_type == "file":
            raise RuntimeError(
                f"GitHub path resolves to a file, not a directory: {owner}/{repo}/{normalized_path}"
            )
        raise RuntimeError(
            f"GitHub path does not resolve to a directory: {owner}/{repo}/{normalized_path}"
        )

    if not isinstance(payload, list):
        raise RuntimeError(
            f"GitHub API returned unexpected data for {owner}/{repo}/{normalized_path}."
        )

    directory_items: list[dict] = []
    for item in payload:
        if not isinstance(item, dict):
            raise RuntimeError(
                f"GitHub directory listing contains unexpected data for {owner}/{repo}/{normalized_path}."
            )
        directory_items.append(item)

    return directory_items


def fetch_languages(owner: str, repo: str, token: Optional[str] = None) -> dict:
    session = _make_session(token)
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/languages"
    return _request_json(session, url)